clang/test/3C/type_params.c

```diff
@@ -157,3 +157,17 @@ void *example1(void * ptr, unsigned int size) {
     // CHECK: ret = realloc<void>(ptr, size);
     return ret;
 }
+
+// Issue #349. Check that the parameter doesn't inherit the double pointer argument within do_doubleptr
+_Itype_for_any(T)
+void incoming_doubleptr(void* ptr : itype(_Array_ptr<T>)) {
+// CHECK_ALL: void incoming_doubleptr(_Array_ptr<T> ptr : itype(_Array_ptr<T>)) {
```

Comment:
Again, this seems very weird to me. I'm not sure why the `void *` component is changing. As far as I am aware, we are assuming that `IType`s will not change during conversion. Perhaps we are slowly moving away from that. @jackastner maybe you can comment?

Reply:
I does seem odd that this ends up with both `_Array_ptr<T>` and `itype(_Array_ptr<T>))`. For the most part, any itypes in the source code should be preserve in the output.

I think the fix this PR makes means that the type of `ptr` is `_Array_ptr<T>` rather than `_Array_ptr<_Ptr<T>>`. This is definitely the better rewriting, but we might want to avoid rewriting the type altogether.

Reply:
Yes, before the fix it was being rewritten with two pointer levels. But it is safe, so shouldn't it be rewritten with the new type and no Itype? Or is that just how it will be done with the liberal itypes enhancement?

Reply:
Two things: (1) It doesn't make sense to have a checked pointer to the left of the colon, on an itype. Having `T: itype(T)` ought to just be `T`. (2) We have not been, by default, rewriting pointers in `itype`s that appear in the source. I think we should stick with this for now. We can consider what changes make sense in a broader issue/PR later, that is disentangled from the issue here (wrong instantiations, and multiple levels of abstraction).

Reply:
Perhaps I misunderstood the bug. In regards to (2), this bugfix does not add the rewriting, it only makes it correct. It was being rewritten in the test originally, but there was no `CHECK` line to verify this. I can look into the rewrite code to try to prevent the rewriting. I assume... anytime `PointerVariableConstraint::ItypeStr` is non-empty?

Reply:
I see. It shouldn't have been getting rewritten, I don't believe. So maybe that's a bug *in addition*? Are these bugs something one can tease apart?

Reply:
If you'll accept separate bugfixes for the two issues mentioned above, than this one as-is fixes the correctness issue, and is ready to merge pending any conflicts with code merged in the mean time. I'll make a new issue for the duplicated type, which is inconvenient but compiles correctly.

Reply:
OK. Just to get it on the record: By "duplicated type" you mean that `void *` is being rewritten to `_Array_ptr<T>` which duplicates the type in `: itype(_Array_ptr<T>)` on the same line. The root type (here, `void *`) should not get rewritten, and that's a separate issue we can tackle (#367).